add flashrom_wp_test_specification.md (#75)
Dasharo/docs: docs/variants/asus_kgpe_d16/dasharo-compatibility/flashrom_wp_test_specification.md

# Flashrom Write Protection – Test Specification

## Test cases

> ### BIOS WP enable test

Reviewer: each test must have its own unique identifier

Reviewer 2: In the test environment, we add a unique identifier: WPR. @PrzemyslawBanasiak13 please add it to test cases

#### Test description

> The test checks if it's possible to set the WP and whether it actually works

Reviewer: missing dot at the end of the sentence

#### Test configuration data

1. `FIRMWARE` = coreboot

2. `OPERATING_SYSTEM` = Ubuntu 20.04

#### Test setup

1. Proceed with the [Generic test setup: firmware](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#firmware)

2. Proceed with the [Generic test setup: OS installer](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-installer)

3. Proceed with the [Generic test setup: OS installation](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-installation)

4. Proceed with the [Generic test setup: OS boot from disk](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-boot-from-disk)

> 5. Check if flashrom and cbfstool are installed:  

Reviewer 2: I think, that we don't need to describe all of the process (checking and installing). We might only write, that the test taker should install flashrom and cbfstool and how to do it.

`$ dpkg -s apache-perl flashrom`  
`$ dpkg -s apache-perl cbfstool`  

6. If flashrom and cbfstool aren't installed, proceed with the below commands:

    - In order to build flashrom we will need some packages and libraries.  
For Debian based distros execute:  
`sudo apt-get install git make binutils build-essential ca-certificates \`  
`libpci-dev libftdi-dev libusb-1.0-0-dev`

    - Now clone the flashrom repository and fetch the patchset:  
`git clone https://github.com/flashrom/flashrom`  
`cd flashrom`  
`git fetch https://review.coreboot.org/flashrom refs/changes/13/59713/7 && \`  
`git checkout FETCH_HEAD`  

    - Build flashrom:  
`make`

7. Ensure the WP pin jumper on the flash is not placed.

#### Test steps

> 1. Clear SPI write protection:  

Reviewer 2: Maybe, if the test step consists of some commands to run in the terminal, we should describe that fact as follows:

1. Clear SPI write protection by using the following commands:

[twice tab] command

If we do it in that way, we will get the effect as in the rest of the documentation.

`./flashrom -p internal --wp-disable`  
`./flashrom -p internal --wp-range=0,0`

2. Check protection status:  
`./flashrom -p internal --wp-status`

3. See what protection ranges are available for the chip:  
`./flashrom -p internal --wp-list`

4. Set protection range e.g.:  
`./flashrom -p internal --wp-range=0x007fc000,0x00004000`  
`./flashrom -p internal --wp-enable`

5. Verify  protection status:
`./flashrom -p internal --wp-status`

#### Expected result

> 1. Flashrom should report:  

Reviewer 2: Same as in the section `Test steps`.

> `Sucessfully set the requested protection range.`, and  

Reviewer: unnecessary comma

`Sucessfully set the requested mode.`

2. WP status should change between 2. and 5. test steps.

### BIOS WP range changing test

#### Test description

The test would check whether it is possible to change the range of the  
protected memory

#### Test configuration data

1. `FIRMWARE` = coreboot

2. `OPERATING_SYSTEM` = Ubuntu 20.04

#### Test setup

1. Proceed with the [Generic test setup: firmware](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#firmware)

2. Proceed with the [Generic test setup: OS installer](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-installer)

3. Proceed with the [Generic test setup: OS installation](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-installation)

4. Proceed with the [Generic test setup: OS boot from disk](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-boot-from-disk)

> 5. Check if flashrom and cbfstool are installed:  

Reviewer 2: Same as in the first test case.

`$ dpkg -s apache-perl flashrom`  
`$ dpkg -s apache-perl cbfstool`  

6. If flashrom and cbfstool aren't installed, proceed with the below commands:  

    - In order to build flashrom we will need some packages and libraries.  
For Debian based distros execute:  
`sudo apt-get install git make binutils build-essential ca-certificates \`  
`libpci-dev libftdi-dev libusb-1.0-0-dev`

    - Now clone the flashrom repository and fetch the patchset:  
`git clone https://github.com/flashrom/flashrom`  
`cd flashrom`  
`git fetch https://review.coreboot.org/flashrom refs/changes/13/59713/7 && \`  
`git checkout FETCH_HEAD`  

    - Build flashrom:  
`make`

#### Test steps

> 1. Check protection status:  

Reviewer 2: Same as in the first test case.

`./flashrom -p internal --wp-status`

2. Disable write protection:  
`./flashrom -p internal --wp-disable`

3. See what protection ranges are available for the chip:  
`./flashrom -p internal --wp-list`

4. Change write protection range e.g.:  
`./flashrom -p internal --wp-range=0x008000,0x00001000`  
`./flashrom -p internal --wp-enable`  

5. Verify protection status:  
`./flashrom -p internal --wp-status`

#### Expected result

> 1. Flashrom should report:  

Reviewer 2: Same as in the first test case.

`Sucessfully set the requested protection range.`, and  
`Sucessfully set the requested mode.`

2. WP status should change between 1. and 5. test steps.

> ### BIOS WP disable test

Reviewer: each test must have its own unique identifier

### Test description

> Test would check whether it is possible to remove WP

Reviewer: keep the description convention (for example from the first test case)

> Test would check whether it is possible to remove WP

Reviewer 2: Same as in the first test case.

#### Test configuration data

1. `FIRMWARE` = coreboot

2. `OPERATING_SYSTEM` = Ubuntu 20.04

#### Test setup

> 1. Proceed with the [Generic test setup: firmware](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#firmware)

Reviewer 2: Same as in the first test case.

2. Proceed with the [Generic test setup: OS installer](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-installer)

3. Proceed with the [Generic test setup: OS installation](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-installation)

4. Proceed with the [Generic test setup: OS boot from disk](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-boot-from-disk)

> 5. Check if flashrom and cbfstool are installed:  

Reviewer: tools

> 5. Check if flashrom and cbfstool are installed:  

Reviewer 2: Same as in the first test case.

`$ dpkg -s apache-perl flashrom`  
`$ dpkg -s apache-perl cbfstool`  

> 6. If flashrom and cbfstool aren't installed, proceed with the below commands:

Reviewer: tools

    - In order to build flashrom we will need some packages and libraries.  
For Debian based distros execute:  
`sudo apt-get install git make binutils build-essential ca-certificates \`  
`libpci-dev libftdi-dev libusb-1.0-0-dev`

    - Now clone the flashrom repository and fetch the patchset:  
`git clone https://github.com/flashrom/flashrom`  
`cd flashrom`  
`git fetch https://review.coreboot.org/flashrom refs/changes/13/59713/7 && \`  
`git checkout FETCH_HEAD`  

    - Build flashrom:  
`make`

> #### Test steps

Reviewer 2: Same as in the first test case.

1. Check protection status:  
`./flashrom -p internal --wp-status`

2. Clear SPI write protection:  
`./flashrom -p internal --wp-disable`  
`./flashrom -p internal --wp-range=0,0`

3. Verify protection status:  
`./flashrom -p internal --wp-status`

#### Expected result

> 1. Command return from step 3. of test steps should include:  

Reviewer 2: Same as in the first test case.

4. Proceed with the [Generic test setup: OS boot from disk](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-boot-from-disk)

5. Install flashrom tools and cbfstool with the below commands:

    - In order to build flashrom we will need some packages and libraries.
For Debian based distros execute:

            `sudo apt-get install git make binutils build-essential ca-certificates \`
            `libpci-dev libftdi-dev libusb-1.0-0-dev`

    - Now clone the flashrom repository and fetch the patchset:

            `git clone https://github.com/flashrom/flashrom`
            `cd flashrom`
            `git fetch https://review.coreboot.org/flashrom refs/changes/13/59713/7 && \`
            `git checkout FETCH_HEAD`

    - Build flashrom:

            `make`

6. Ensure the WP pin jumper on the flash is not placed.

#### Test steps

1. Clear SPI write protection:

        `./flashrom -p internal --wp-disable`
        `./flashrom -p internal --wp-range=0,0`

2. Check protection status:

        `./flashrom -p internal --wp-status`

3. See what protection ranges are available for the chip:

        `./flashrom -p internal --wp-list`

> 4. Set protection range e.g.:

Reviewer 2: At this point we should add reference to document in which we describe how to choose the correct write protection range

        `./flashrom -p internal --wp-range=0x007fc000,0x00004000`
        `./flashrom -p internal --wp-enable`

5. Verify  protection status:

        `./flashrom -p internal --wp-status`

#### Expected result

1. Flashrom should report:

        `Sucessfully set the requested protection range.`
    and

        `Sucessfully set the requested mode.`

> 2. WP status should change between 2. and 5. test steps.

Reviewer 2: At this point, it would be good if we show output sample

4. Proceed with the [Generic test setup: OS boot from disk](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-boot-from-disk)

5. Install flashrom tools and cbfstool with the below commands with the below commands:

    - In order to build flashrom we will need some packages and libraries.
For Debian based distros execute:

            `sudo apt-get install git make binutils build-essential ca-certificates \`
            `libpci-dev libftdi-dev libusb-1.0-0-dev`

    - Now clone the flashrom repository and fetch the patchset:

            `git clone https://github.com/flashrom/flashrom`
            `cd flashrom`
            `git fetch https://review.coreboot.org/flashrom refs/changes/13/59713/7 && \`
            `git checkout FETCH_HEAD`

    - Build flashrom:

            `make`

#### Test steps

1. Check protection status:

        `./flashrom -p internal --wp-status`

2. Disable write protection:

        `./flashrom -p internal --wp-disable`

3. See what protection ranges are available for the chip:

        `./flashrom -p internal --wp-list`

> 4. Change write protection range e.g.:

Reviewer 2: At this point we should add reference to document in which we describe how to choose the correct write protection range

# Flashrom Write Protection – Test Specification

## Test cases

> ### WPR001.001 BIOS WP enable test

Reviewer: please delete `test` in the title

        `./flashrom -p internal --wp-range=0,0`

2. Check protection status:

        `./flashrom -p internal --wp-status`

3. See what protection ranges are available for the chip:

        `./flashrom -p internal --wp-list`

4. Set protection range e.g.(details on WP ranges [here](https://docs.dasharo.com/variants/asus_kgpe_d16/spi-wp/)):

        `./flashrom -p internal --wp-range=0x007fc000,0x00004000`
        `./flashrom -p internal --wp-enable`

5. Verify  protection status:

        `./flashrom -p internal --wp-status`

#### Expected result

1. Flashrom should report:

        `Sucessfully set the requested protection range.`
    and

        `Sucessfully set the requested mode.`

2. WP status should change between 2. and 5. test steps. Output samples:

    Step 2:

    ```
    WP config bits: SRP1=0 SRP0=0 CMP=0 SEC=0 TB=0 BP2=0 BP1=0 BP0=0
    Protection range: start=0x00000000 length=0x00000000 (none)
    Protection mode: disabled
    ```

    Step 5:

    ```
    WP config bits: SRP1=0 SRP0=1 CMP=0 SEC=1 TB=0 BP2=0 BP1=1 BP0=1
    Protection range: start=0x007fc000 length=0x00004000 (upper 1/512)
    Protection mode: hardware
    ```

> ### WPR002.001 BIOS WP range changing test

Reviewer: please delete test in the title

#### Test description

This test aims to verify whether it is possible to change the range of the

> protected memory

Reviewer: missing dot at the end of the sentence

Reviewer: @PrzemyslawBanasiak13 ping

#### Test configuration data

1. `FIRMWARE` = coreboot

2. `OPERATING_SYSTEM` = Ubuntu 20.04

#### Test setup

1. Proceed with the [Generic test setup: firmware](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#firmware)

2. Proceed with the [Generic test setup: OS installer](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-installer)

3. Proceed with the [Generic test setup: OS installation](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-installation)

4. Proceed with the [Generic test setup: OS boot from disk](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-boot-from-disk)

> 5. Install flashrom tools and cbfstool with the below commands with the below commands:

Reviewer: `flashrom and cbftool tools`

5. Verify protection status:

        `./flashrom -p internal --wp-status`

#### Expected result

1. Flashrom should report:

        `Sucessfully set the requested protection range.`
    and

        `Sucessfully set the requested mode.`
2. WP status should change between 1. and 5. test steps.

### BIOS WP disable test

### Test description

This test aims to verify whether it is possible to remove write protecion.

#### Test configuration data

1. `FIRMWARE` = coreboot

2. `OPERATING_SYSTEM` = Ubuntu 20.04

#### Test setup

1. Proceed with the [Generic test setup: firmware](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#firmware)

2. Proceed with the [Generic test setup: OS installer](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-installer)

3. Proceed with the [Generic test setup: OS installation](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-installation)

4. Proceed with the [Generic test setup: OS boot from disk](https://novacustom.gitlab.io/dasharo-compatibility/dasharo-compatibility/generic-test-setup/#os-boot-from-disk)

> 5. Install flashrom tools and cbfstool with the below commands with the below commands:

Reviewer: `flashrom and cbftool tools`

    - In order to build flashrom we will need some packages and libraries.
For Debian based distros execute:

            `sudo apt-get install git make binutils build-essential ca-certificates \`
            `libpci-dev libftdi-dev libusb-1.0-0-dev`

    - Now clone the flashrom repository and fetch the patchset:

            `git clone https://github.com/flashrom/flashrom`
            `cd flashrom`
            `git fetch https://review.coreboot.org/flashrom refs/changes/13/59713/7 && \`
            `git checkout FETCH_HEAD`

    - Build flashrom:

            `make`

#### Test steps

1. Check protection status:

        `./flashrom -p internal --wp-status`

2. Clear SPI write protection:

        `./flashrom -p internal --wp-disable`
        `./flashrom -p internal --wp-range=0,0`

3. Verify protection status:

        `./flashrom -p internal --wp-status`

#### Expected result

> 1. Command return from step 3. of test steps should include:

Reviewer: ```suggestion
1. Command output from step 3. of test steps should be similar to:
```
